metrics_utility/test/ccspv_reports/dedup/test_complex_CCSPv2_with_canonical_facts/test_complex_CCSPv2_with_canonical_facts.py

```diff
@@ -0,0 +1,2459 @@
+         - Same VMware serial + machine_id (01b6b28643a6a867e339e957c8ed9d37)
+         - Production variants of same web server
+         - Result: Merged into web03.internal
+         - Dedup: New logic applied (count=2) - both had different ansible_host and different host_name
+
+    1.6. cache01.internal (2 entries → 1):
+         - Both have same machine_id (0267fc0887de14e8c994d1025a445221) but NO product_serial
+         - From different orgs (Production, Development)
+         - Result: Merged because machine_id matches (serial not required if missing)
+         - Dedup: Old logic only (count=1) - same ansible_host and host_name
+
+    2. NOT DEDUPLICATED HOSTS (unique serial/machine_id combinations):
+    -------------------------------------------------------------------
+
+    2.1. db01.company.com:
+         - Has product_serial but NO machine_id
+         - Cannot deduplicate without machine_id
+         - Result: Kept separate
+         - Dedup: No dedup needed (count=1) - unique ansible_host and serial CN7792194B0740
+
+    2.2. log01.company.com:
+         - Missing BOTH product_serial AND machine_id
+         - No canonical facts to deduplicate on
+         - Result: Kept separate
+         - Dedup: No dedup needed (count=1) - unique host
+
+    2.3. web04.dev and web04.staging:
+         - Different machine_ids (ae920ed940e880003e264a357de969c1 vs d1134fec21d571a9b596f7dbf7dc5673)
+         - Different serials (VMware-dev-... vs VMware-stg-...)
+         - Different hostnames
+         - Result: Kept as separate hosts (different environments)
+         - Dedup: No dedup needed (count=1 each) - different hosts
+
+    3. FALSE NEGATIVES - NOT DEDUPLICATED (but should be):
+    -------------------------------------------------------
+    3.1. win-srv01.company.com and win-srv02.company.com:
+         - Different Windows servers with SAME serial (USE9876543)
+         - Windows lacks machine_id (systemd-specific)
+         - Only product_serial available for deduplication
+         - Result: Kept separate (FALSE NEGATIVE - same serial but no machine_id)
+         - Dedup: No dedup applied (count=1 each) - new logic requires machine_id
+
+    3.2. k8s-node-01.cluster and k8s-node-01.internal:
+         - Same Kubernetes node accessed differently
+         - Container environment lacks both machine_id and serial
+         - No canonical facts for deduplication
+         - Result: Kept separate (SHOULD be merged based on hostname pattern)
+         - Dedup: No dedup possible - no canonical facts
+
+    3.3. secure-host-01.company.com and secure-host-01-readonly.internal (privileged vs unprivileged):
+         - Same host accessed with different credentials
+         - Admin job has product_serial, user job doesn't
+         - Same machine_id in both cases 4f7a8b9c2d3e5f6a7b8c9d0e1f2a3b4c
+         - Result: Correctly merged based on machine_id
+         - Dedup: Working correctly - same machine_id causes deduplication
```

Comment:
this is wrong - if the user job doesn't have product serial, and uses different hostnames, it won't get merged.
We don't merge on machine_id alone when product_serial is missing.

Reply:
ah, yes, it's just a description error (they are correctly in false negatives category)